// azure-cli/lib/commands/arm/hdinsight/hdiClient.js
/*** Generated by streamline 0.10.17 (callbacks) - DO NOT EDIT ***/ var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename, false),__func=__rt.__func,__cb=__rt.__cb,__catch=__rt.__catch,__tryCatch=__rt.__tryCatch; var __ = require("underscore");
















var profile = require("../../../util/profile");
var utils = require("../../../util/utils");
var HdiCustomization = require("./hdiCustomization");
var HdiConfigClient = require("./hdiConfigClient");

function HdiClient(cli, subscription) {
  this.cli = cli;
  this.subscription = subscription;};


__.extend(HdiClient.prototype, {
  createCluster: function createCluster__1(resourceGroupName, clusterName, clusterCreateParameters, _) { var customization, result, __this = this; var __frame = { name: "createCluster__1", line: 29 }; return __func(_, this, arguments, createCluster__1, 3, __frame, function __$createCluster__1() {
      customization = new HdiCustomization(__this.cli);
      return customization.createCluster(resourceGroupName, clusterName, clusterCreateParameters, __cb(_, __frame, 2, 31, function ___(__0, __1) { result = __1;
        return _(null, result); }, true)); }); },


  getCluster: function getCluster__2(resourceGroupName, clusterName, _) { var subscription, client, __this = this; var __frame = { name: "getCluster__2", line: 35 }; return __func(_, this, arguments, getCluster__2, 2, __frame, function __$getCluster__2() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$getCluster__2() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription);
            return client.clusters.get(resourceGroupName, clusterName, __cb(_, __frame, 4, 29, _, true)); }); })(function ___(e, __result) { __catch(function __$getCluster__2() { if (e) {


              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$getCluster__2() { _(); }); }); }); },



  listClusters: function listClusters__3(resourceGroupName, _) { var subscription, client, __this = this; var __frame = { name: "listClusters__3", line: 49 }; return __func(_, this, arguments, listClusters__3, 1, __frame, function __$listClusters__3() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$listClusters__3() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription); return (function __$listClusters__3(__then) {
              if (((resourceGroupName !== undefined) && (resourceGroupName !== null))) {
                return client.clusters.listByResourceGroup(resourceGroupName, __cb(_, __frame, 5, 31, _, true)); } else { __then(); } ; })(function __$listClusters__3() {

              return client.clusters.list(__cb(_, __frame, 7, 29, _, true)); }); }); })(function ___(e, __result) { __catch(function __$listClusters__3() { if (e) {


              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$listClusters__3() { _(); }); }); }); },



  deleteCluster: function deleteCluster__4(resourceGroupName, clusterName, _) { var subscription, client, __this = this; var __frame = { name: "deleteCluster__4", line: 66 }; return __func(_, this, arguments, deleteCluster__4, 2, __frame, function __$deleteCluster__4() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$deleteCluster__4() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription);
            return client.clusters.deleteMethod(resourceGroupName, clusterName, __cb(_, __frame, 4, 29, _, true)); }); })(function ___(e, __result) { __catch(function __$deleteCluster__4() { if (e) {

              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$deleteCluster__4() { _(); }); }); }); },



  resizeCluster: function resizeCluster__5(resourceGroupName, clusterName, targetInstanceCount, _) { var subscription, client, resizeParameter, __this = this; var __frame = { name: "resizeCluster__5", line: 79 }; return __func(_, this, arguments, resizeCluster__5, 3, __frame, function __$resizeCluster__5() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$resizeCluster__5() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription);
            resizeParameter = {
              targetInstanceCount: targetInstanceCount };

            return client.clusters.resize(resourceGroupName, clusterName, resizeParameter, __cb(_, __frame, 7, 29, _, true)); }); })(function ___(e, __result) { __catch(function __$resizeCluster__5() { if (e) {

              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$resizeCluster__5() { _(); }); }); }); },



  enableHttp: function enableHttp__6(resourceGroupName, clusterName, userName, password, _) { var subscription, client, httpSettingsParameters, __this = this; var __frame = { name: "enableHttp__6", line: 95 }; return __func(_, this, arguments, enableHttp__6, 4, __frame, function __$enableHttp__6() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$enableHttp__6() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription);
            httpSettingsParameters = {
              httpUserEnabled: true,
              httpUsername: userName,
              httpPassword: password };

            return client.clusters.configureHttpSettings(resourceGroupName, clusterName, httpSettingsParameters, __cb(_, __frame, 9, 29, _, true)); }); })(function ___(e, __result) { __catch(function __$enableHttp__6() { if (e) {

              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$enableHttp__6() { _(); }); }); }); },



  disableHttp: function disableHttp__7(resourceGroupName, clusterName, _) { var subscription, client, httpSettingsParameters, __this = this; var __frame = { name: "disableHttp__7", line: 113 }; return __func(_, this, arguments, disableHttp__7, 2, __frame, function __$disableHttp__7() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$disableHttp__7() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription);
            httpSettingsParameters = {
              httpUserEnabled: false };

            return client.clusters.configureHttpSettings(resourceGroupName, clusterName, httpSettingsParameters, __cb(_, __frame, 7, 29, _, true)); }); })(function ___(e, __result) { __catch(function __$disableHttp__7() { if (e) {

              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$disableHttp__7() { _(); }); }); }); },



  enableRdp: function enableRdp__8(resourceGroupName, clusterName, rdpUserName, rdpPassword, rdpExpiryDate, _) { var subscription, client, rdpSettingsParameters, __this = this; var __frame = { name: "enableRdp__8", line: 129 }; return __func(_, this, arguments, enableRdp__8, 5, __frame, function __$enableRdp__8() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$enableRdp__8() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription);
            rdpSettingsParameters = {
              osProfile: {
                windowsOperatingSystemProfile: {
                  rdpSettings: {
                    userName: rdpUserName,
                    password: rdpPassword,
                    expiryDate: rdpExpiryDate } } } };




            return client.clusters.configureRdpSettings(resourceGroupName, clusterName, rdpSettingsParameters, __cb(_, __frame, 15, 29, _, true)); }); })(function ___(e, __result) { __catch(function __$enableRdp__8() { if (e) {

              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$enableRdp__8() { _(); }); }); }); },



  disableRdp: function disableRdp__9(resourceGroupName, clusterName, _) { var subscription, client, rdpSettingsParameters, __this = this; var __frame = { name: "disableRdp__9", line: 153 }; return __func(_, this, arguments, disableRdp__9, 2, __frame, function __$disableRdp__9() { return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$disableRdp__9() {

            subscription = profile.current.getSubscription(__this.subscription);
            client = utils.createHDInsightManagementClient(subscription);
            rdpSettingsParameters = {
              osProfile: {
                windowsOperatingSystemProfile: {
                  rdpSettings: null } } };



            return client.clusters.configureRdpSettings(resourceGroupName, clusterName, rdpSettingsParameters, __cb(_, __frame, 11, 29, _, true)); }); })(function ___(e, __result) { __catch(function __$disableRdp__9() { if (e) {

              if (((e.code === "ResourceNotFound") || (e.code === "NotFound"))) {
                return _(null, null); } ;

              return _(e); } else { _(null, __result); } ; }, _); }); })(function ___() { __tryCatch(_, function __$disableRdp__9() { _(); }); }); }); },



  createConfigFile: function(configFilePath) {
    var configClient = new HdiConfigClient(this.cli);
    var result = configClient.createConfigFile(configFilePath);
    return result; },


  addConfigValue: function(configFilePath, options) {
    var configClient = new HdiConfigClient(this.cli);
    var result = configClient.addConfigValue(configFilePath, options);
    return result; },


  addScriptAction: function(configFilePath, options) {
    var configClient = new HdiConfigClient(this.cli);
    var result = configClient.addScriptAction(configFilePath, options);
    return result; }});



module.exports = HdiClient;
